src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLambdaParameterShouldBeUnnamedCheck.java

```diff
@@ -0,0 +1,284 @@
+// checkstyle: Checks Java source code and other text files for adherence to a set of rules.
+// Copyright (C) 2001-2024 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Optional;
+
+import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Ensures that lambda parameters that are not used are declared as an unnamed variable.
+ * </p>
+ * <p>
+ * Rationale:
+ * </p>
+ * <ul>
+ *     <li>
+ *         Improves code readability by clearly indicating which parameters are unused.
+ *     </li>
+ *     <li>
+ *         Follows Java conventions for denoting unused parameters with an underscore ({@code _}).
+ *     </li>
+ * </ul>
+ * <p>
+ * See the <a href="https://docs.oracle.com/en/java/javase/21/docs/specs/unnamed-jls.html">
+ * Java Language Specification</a> for more information about unnamed variables.
+ * </p>
+ * <p>
+ * <b>Attention</b>: This check should be activated only on source code
+ * that is compiled by jdk21 or higher;
+ * unnamed lambda parameters came out as the first preview in Java 21.
+ * </p>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code unused.lambda.parameter}
+ * </li>
+ * </ul>
+ *
+ * @since 10.18.0
+ *
+ */
+@FileStatefulCheck
+public class UnusedLambdaParameterShouldBeUnnamedCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_UNUSED_LAMBDA_PARAMETER = "unused.lambda.parameter";
+
+    /**
+     * Invalid parents of the lambda parameter identifier.
+     */
+    private static final int[] INVALID_LAMBDA_PARAM_IDENT_PARENTS = {
```

Comment:
To ignore identifiers when they are method call or type etc..

examples: https://github.com/checkstyle/checkstyle/blob/ca17d061268c738ce72e24c441cea56bec0f2a94/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unusedlambdaparametershouldbeunnamed/InputUnusedLambdaParameterShouldBeUnnamedMultipleParameters.java#L38-L50

line 40 is type, line 49 is method call so we should ignore this and not be considered as candidate for lambda parameter ident

same as: https://github.com/checkstyle/checkstyle/pull/15358#discussion_r1691996520

Reply:
>same as: https://github.com/checkstyle/checkstyle/pull/15358#discussion_r1691996520

Yes, yet another ambiguously named collection. 

>line 40 is type, line 49 is method call so we should ignore this and not be considered as candidate for lambda parameter ident

>INVALID_LAMBDA_PARAM_IDENT_PARENTS

This doesn't add up. Why are we calling this LAMBDA_PARAM anything if the identifiers they are on are not lambda parameters?

Imagine you did not write this check, and need to fix some bug. It is not clear how these tokens are related, and what "invalid" means in this context. Instead of explicitly mentioning what a lambda parameter should never have as a parent, why not just call out exactly what CAN be a parent? This should be simple to figure out from our grammar/visitor.

Reply:
> Why are we calling this LAMBDA_PARAM anything if the identifiers they are on are not lambda parameters?

that's why we perfixed with "INVALID"

> Instead of explicitly mentioning what a lambda parameter should never have as a parent, why not just call out exactly what CAN be a parent?

I assumed that this would be a huge list so it is better to have the opposite (tokens that CAN NOT be parent)

what about adding this extra explanation in Javadoc:  `These are tokens that can not be parents for a lambda
 parameter identifier`
 
does it become less ambiguous, IYO?

Reply:
This is better for sure.